_inc/search/components/search-widget.jsx

```diff
@@ -0,0 +1,78 @@
+/** @jsx h */
+
+/**
+ * External dependencies
+ */
+import { h, Component } from 'preact';
+import Portal from 'preact-portal';
+import { debounce } from 'lodash';
+
+class SearchWidget extends Component {
+	constructor() {
+		super( ...arguments );
+		this.requestId = 0;
+		this.state = {
+			query: this.props.initialValue,
+			results: [],
+		};
+		this.onChangeQuery = this.onChangeQuery.bind( this );
+		this.getResults = debounce( this.getResults.bind( this ), 500 );
+		this.getResults( this.props.initialValue );
+	}
+	componentDidMount() {
+		if ( this.props.grabFocus ) {
+			this.input.focus();
+		}
+	}
+
+	bindInput = input => ( this.input = input );
+	onChangeQuery = event => {
+		const query = event.target.value;
+		this.setState( { query } );
+		this.getResults( query );
+	};
+
+	getResults( query ) {
+		if ( query ) {
+			const { api } = this.props;
+			if ( api ) {
+				this.requestId++;
+				const requestId = this.requestId;
+				api
+					.fetch( query )
+					.then( response => {
+						return response.json();
+					} )
+					.then( json => {
+						if ( this.requestId === requestId ) {
+							this.setState( { results: json } );
+						}
+					} );
+			}
+		} else {
+			this.setState( { results: [] } );
+		}
+	}
+
+	render() {
+		const { query, results } = this.state;
+		const { SearchResults, api } = this.props;
+		return (
+			<div>
+				<p>
+					<input
+						type="text"
+						value={ query }
+						onInput={ this.onChangeQuery }
+						ref={ this.bindInput }
+					/>
+				</p>
+				<Portal into="#results">
```

Comment:
You'll prolly need to change this ID attribute to something more unique.

Reply:
Why do we need a `Portal` here? (Can you maybe provide a screenshot to illustrate it?) Typically Portals should be reserved to control spatially disjoint pieces of the UI -- like a Gutenberg block controlling the elements in the editor sidebar (but we have higher-level interfaces for cases like that.)

I'd like us to be super-cautious about introducing `Portal`s -- since they allow us to break out of React's typical component hierarchy, they come with a risk of :spaghetti: UI code.

Reply:
Potential use case here is the search widget is in the sidebar or header, but the results appear in the main content area.

Reply:
That makes sense :thinking: Curious about some UI and implementation aspects though -- would the user add the widget to the sidebar, and a results block to a post's content? Or to a theme/page template? The [ID problem](https://github.com/Automattic/jetpack/pull/12741#discussion_r299356870) @simison has pointed out possibly highlights some conceptual issues with this. (E.g. I guess we can restrict the results block to appear only once in a post, but there could still be multiple posts with that block, visible at the same time in an archive-type page.) Maybe that's something complex enough to defer to a future iteration?

Reply:
p3QzjZ-VL has mockups which hopefully clarifies some of this. I'm unsure if this needs an additional summary added to #13027 or not, but am pretty sure all of this code will change as we get more into the implementation.